cmd/frontman/main.go

```diff
@@ -137,245 +165,315 @@ sudo sysctl -w net.ipv4.ping_group_range="0   2147483647"`
+	case "unix-systemv":
+		fmt.Printf("Run this command to stop it:\nsudo service %s stop\n\n", svcConfig.Name)
+	case "linux-upstart":
+		fmt.Printf("Run this command to stop it:\nsudo initctl stop %s\n\n", svcConfig.Name)
+	case "linux-systemd":
+		fmt.Printf("Run this command to stop it:\nsudo systemctl stop %s.service\n\n", svcConfig.Name)
+	case "darwin-launchd":
+		fmt.Printf("Run this command to stop it:\nsudo launchctl unload %s\n\n", svcConfig.Name)
+	case "windows-service":
+		fmt.Printf("Use the Windows Service Manager to stop it\n\n")
+	}
 
-			if err != nil && strings.Contains(err.Error(), "already exists") {
+	fmt.Printf("Log file located at: %s\n", fm.LogFile)
+	os.Exit(0)
+}
 
-				fmt.Printf("Frontman service(%s) already installed: %s\n", systemManager.String(), err.Error())
+func runUnderOsServiceManager(fm *frontman.Frontman) {
+	systemService, err := getServiceFromFlags(fm, "", "")
+	if err != nil {
+		log.Fatal(err)
+	}
 
-				note := ""
-				if runtime.GOOS == "windows" {
-					note = " Windows Services Manager app should not be opened!"
-				}
-				if askForConfirmation("Do you want to overwrite it?" + note) {
-					s.Stop()
-					err := s.Uninstall()
-					if err != nil {
-						fmt.Printf("Failed to unistall the service: %s\n", err.Error())
-						return
-					}
-					goto install
-				}
-				s.Uninstall()
+	// we are running under OS service manager
+	err = systemService.Run()
 
-			} else if err != nil {
-				fmt.Printf("Frontman service(%s) installing error: %s", systemManager.String(), err.Error())
-				return
-			} else {
-				fmt.Printf("Frontman service(%s) installed. Starting...\n", systemManager.String())
-			}
+	if err != nil {
+		log.Fatal(err.Error())
+	}
 
-			err = s.Start()
+	os.Exit(0)
+}
 
-			if err != nil {
-				fmt.Printf("Already running\n")
-			}
+// serviceWrapper is created to provide context and methods that satisfies service.Interface
+// in order to run Frontman under OS Service Manager
+type serviceWrapper struct {
+	Frontman      *frontman.Frontman
+	InterruptChan chan struct{}
+	DoneChan      chan struct{}
+}
 
-			switch systemManager.String() {
-			case "unix-systemv":
-				fmt.Printf("Run this command to stop/start it:\nsudo service %s stop\nsudo service %s start\n\n", svcConfig.Name, svcConfig.Name)
-			case "linux-upstart":
-				fmt.Printf("Run this command to stop/start it:\nsudo initctl stop %s\nsudo initctl start %s\n\n", svcConfig.Name, svcConfig.Name)
-			case "linux-systemd":
-				fmt.Printf("Run this command to stop/start it:\nsudo systemctl stop %s.service\nsudo systemctl start %s.service\n\n", svcConfig.Name, svcConfig.Name)
-			case "darwin-launchd":
-				fmt.Printf("Run this command to stop/start it:\nsudo launchctl unload %s\nsudo launchctl load /Library/LaunchDaemons/%s.plist\n\n", svcConfig.Name, svcConfig.Name)
-			case "windows-service":
-				fmt.Printf("Use the Windows Service Manager to stop/start it\n\n")
-			}
+func (sw *serviceWrapper) Start(s service.Service) error {
```

Comment:
why have return type `error` if there is no case where an error is returned? Or does this come from `service.Interface`?

Reply:
Yes, it should implement `service.Interface`